### DISCUSSION TITLE
[1일차-윤경]가상돔이란?

### DISCUSSION THREAD

**Discussion by skyhanull** [zzangCS/zzangInterview#6] [프론트 면접 질문]
react는 뒤에서 실제 DOM을 모두 복사해서 virtual DOM을 만든다.

만약 리액트를 이용해서 virtual dom의 어떤 노드를 수정한다면 리액트는 새로운 virtual dom(updated virtual dom)을 만들고 이전 virtual dom 트리와 비교한다.

만약 일치하지 않는 노드를 발견한다면 리액트는 실제 dom 트리에서 해당 노드만 업데이트 시켜준다.

**Comment by ez1n** [comment]
가상 돔 (Virtual DOM)이란 View를 실제 DOM이 아닌 Virtual DOM과 메모리에서 미리 처리하고 저장한 후 실제 DOM과 동기화하는 프로그램 개념으로 실제 DOM보다 가볍고 빠른 렌더링이 가능합니다.

특정 페이지에서 데이터 변화가 일어났을 때 새로운 상태를 Virtual DOM으로 만든 뒤 기존 Virtual DOM과 비교하여 변경이 있는 부분만 실제 DOM에 적용하여 업데이트합니다.

**Reply by ez1n** [reply]
- **DOM (Document Object Model)**
    
    ⇒ HTML과 XML 문서의 프로그래밍 인터페이스 (nodes, objects로 문서 표현)
    
    ⇒ HTML은 브라우저에서 실행도리 수 있게끔 DOM Tree로 파싱되고 이를 바탕으로 렌더링한다.
    
    ⇒ 문서의 구조화된 표현을 제공하며 프로그래밍 언어가 DOM 구조에 접근할 수 있는 방법을 제공하여 문서 구조, 스타일, 내용 등을 변경할 수 있게 돕는 역할
    
    ⇒ web의 복잡도가 증가함에 따라 DOM 조작이 빈번하게 발생하는 것을 해결하기 위해 가상 돔 등장 (DOM의 조작은 브라우저 렌더 과정을 유발한다. → 비효율적!)
    
- **V-DOM (Virtual DOM)**
    
    ⇒ DOM을 직접 조작하지 않고 변경사항을 하나의 V-DOM에 모았다가 DOM에 한 번에 보내는 기술
    
    ⇒ 하나의 가상 레이어로 V-DOM에 적용하는 변경사항은 DOM에 바로 적용되지 않음
    
    ⇒ 변경 사항들을 모아뒀다가 한 번만 업데이트(렌더링) 하기 때문에 DOM 조작이 빈번할 경우 효율적으로 동작함
    
    ⇒ DOM의 조작이 빈번하지 않은 경우 DOM을 직접 조작하는 것보다 느리다.
    

- **React(리액트)의 등장**
    
    ⇒ 데이터가 바뀔 때마다 DOM을 새로 그리자!
    
    ⇒ V-DOM에 새로 그리고 실제 DOM과 비교하여 변경된 부분만 한 번에 보내는 방법을 사용
    
    ⇒ 데이터의 흐름과 DOM의 차이에 의한 버그 발생 X
    
    ⇒ 상태에만 집중하여 선언적인 코드를 짤 수 있다!

**Comment by smarfy99** [comment]
가상 DOM이란 실제 돔을 추상화한 JS 객체이다.
DOM을 변경하려면 매번 모든 document 객체를 훑고, 변경된 돔을 다시 그려주는데 메모리 누수가 발생한다. 가상 돔은 돔을 추상화해 변화된 부분만 가상돔에 적용시켜 가상돔끼리의 비교로 연산 시간을 줄이고, 메모리 누수를 방지한다.

**Reply by smarfy99** [reply]
https://velog.io/@mollog/React%EC%97%90%EC%84%9C%EC%9D%98-%EA%B0%80%EC%83%81%EB%8F%94-%EA%B0%9C%EB%85%90
